### PR TITLE
Improves accountExists method to discern error codes

### PR DESCRIPTION
**Pull request by amiecorso**
## High Level Overview of Change
accountExists method in XpringClient was treating any exception raised as a "false" conclusion (account does not exist).  This PR updates the method to conclude false only if a grpc.status.NOT_FOUND error code is in the exception.  Other codes will cause the exception to be re-thrown.  Unit and integration tests also updated to test ability to distinguish error codes.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (non-breaking change that only restructures code)

## Before / After
Before: accountExists could have returned misleading false-negatives. 
After: will propagate exceptions that it's not qualified to interpret.

## Test Plan
CI

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -18,5 +18,11 @@
     },
     "files.insertFinalNewline": true,
     "files.trimFinalNewlines": true,
-    "files.trimTrailingWhitespace": true
+    "files.trimTrailingWhitespace": true,
+    "mocha.options": {},
+    "mocha.requires": [
+        "ts-node/register"
+    ],
+    "mocha.files.glob": "test/**/*.js"
```
Comment (Stormtv) [inline .vscode/settings.json]: This should probably be `"test/**/*-test.ts"`

File: test/mocha.opts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+--require ./node_modules/ts-node/register	
+--require ./node_modules/source-map-support/register	
+--recursive	
+test/**/*.ts	
+
```
Comment (Stormtv) [inline test/mocha.opts]: Mocha opts is deprecated and was replaced by the .mocharc.js

https://mochajs.org/#mochaopts

File: src/default-xpring-client.ts [hunk 3]
```diff
@@ -299,7 +300,10 @@ class DefaultXpringClient implements XpringClientDecorator {
       await this.getBalance(address)
       return true
     } catch (e) {
-      return false
+      if ('code' in e && e.code === grpc.status.NOT_FOUND) {
```
Comment (Stormtv) [inline src/default-xpring-client.ts]: I believe best practice here might be to use `hasOwnProperty` vs `in`

File: src/legacy/legacy-default-xpring-client.ts [hunk 4]
```diff
@@ -244,7 +245,10 @@ class LegacyDefaultXpringClient implements XpringClientDecorator {
       await this.getBalance(address)
       return true
     } catch (e) {
-      return false
+      if ('code' in e && e.code === grpc.status.NOT_FOUND) {
```
Comment (Stormtv) [inline src/legacy/legacy-default-xpring-client.ts]: I believe best practice here might be to use `hasOwnProperty` vs `in`

File: .mocharc.js [hunk 5]
```diff
@@ -1,23 +0,0 @@
...
-
-  // Warn if test exceed 75ms duration
-  slow: 75,
-
-  // Fail if tests exceed 10000ms
-  timeout: 10000,
-
-  // Check for global variable leaks
-  'check-leaks': true,
-}
```
Comment (Stormtv) [inline .mocharc.js]: Why was this deleted?

File: .vscode/settings.json (hunk 1 above)
Comment (amiecorso) [reply .vscode/settings.json]: Cool! Do you mind explaining what this line is for?  I think it may have been auto-generated because I definitely didn't write it :)

File: .mocharc.js (hunk 5 above)
Comment (amiecorso) [reply .mocharc.js]: hmm, I also don't know.  Also something I didn't touch.

File: src/legacy/legacy-default-xpring-client.ts (hunk 4 above)
Comment (amiecorso) [reply src/legacy/legacy-default-xpring-client.ts]: great, I was actually wondering which one was better so I'll change that.

Comment (amiecorso) [reply src/legacy/legacy-default-xpring-client.ts]: So I just pushed that change  - the linter wanted me to write it like this: Object.prototype.hasOwnProperty.call(e, 'code')   instead of just e.hasOwnProperty('code').  Is that ok?

Comment (Stormtv) [reply src/legacy/legacy-default-xpring-client.ts]: Yeah, thats perfect 👍

File: .vscode/settings.json (hunk 1 above)
Comment (Stormtv) [reply .vscode/settings.json]: This is the workspace mocha configuration for the mocha test panel extension in VSCode.

`test/**/*.js` is telling it to look for files of the extension `.js` in a folder called test and all the folders underneath it. 

Our tests are written in Typescript so they have the `.ts` extension and they always follow the executable tests always follow the format of `name-test.ts`.

Glob patterns are interesting there is a great resource at https://globster.xyz/ to practice and see how they work

**Comment by Stormtv** [comment]
I changed the branch this PR is compared to as I see now that #182 is what you were building on top of which is where those mocha changes came in.

File: test/mocha.opts (hunk 2 above)
Comment (keefertaylor) [reply test/mocha.opts]: Agree this change shouldn't be part of this PR. 

This change broke VSCode's test explorer and I'm not sure why.  I'm rolling it back here: https://github.com/xpring-eng/Xpring-JS/pull/182

File: test/default-xpring-client-test.ts [hunk 6]
```diff
@@ -32,6 +33,23 @@ const fakeErroringNetworkClient = new FakeNetworkClient(
   FakeNetworkClientResponses.defaultErrorResponses,
 )
 
+/**
+ * Convenience class for creating more specific Error objects to mimick grpc error responses.
+ */
+class FakeServiceError extends Error {
```
Comment (keefertaylor) [inline test/default-xpring-client-test.ts]: Maybe call this FakeGRPCError since it's mirroring gRPC semantics?

File: test/default-xpring-client-test.ts [hunk 7]
```diff
@@ -32,6 +33,23 @@ const fakeErroringNetworkClient = new FakeNetworkClient(
...
+  /**
+   * Construct a new FakeServiceError.
+   *
+   * @param message The text details of the error.
+   * @param code The grpc.status code.
+   */
+  constructor(public message: string, public code: grpc.status) {
```
Comment (keefertaylor) [inline test/default-xpring-client-test.ts]: I think that lines 48/49 are redundant (they get automatically added by putting a `public` modifier in front of these properties.

Could you also make them `readonly`? It doesn't really matter, but immutability is a good habit.

File: test/default-xpring-client-test.ts [hunk 8]
```diff
@@ -436,17 +454,52 @@ describe('Default Xpring Client', function(): void {
...
+    const fakeNetworkClientResponses = new FakeNetworkClientResponses(
+      errorWithCode5, // getAccountInfoResponse
+    )
+    const fakeErroringNetworkClient = new FakeNetworkClient(
+      fakeNetworkClientResponses,
+    )
     const xpringClient = new DefaultXpringClient(fakeErroringNetworkClient)
 
-    // WHEN accountExists throws an exception while calling getBalance
+    // WHEN accountExists encounters this Error while calling getBalance
```
Comment (keefertaylor) [inline test/default-xpring-client-test.ts]: Generally, encourage you to think of GIVEN, WHEN, THEN as a human reading it, rather than just writing what the test is doing.

You could make this as simple as:
// GIVEN A xpring client with a network client that will report accounts as not found
// WHEN Account existence is checked
// THEN The account is reported to exist

File: test/default-xpring-client-test.ts [hunk 9]
```diff
@@ -436,17 +454,52 @@ describe('Default Xpring Client', function(): void {
     assert.equal(exists, true)
   })
 
-  it('Check if account exists - failing network request', async function() {
-    // GIVEN a XpringClient which wraps an erroring network client.
+  it('Check if account exists - failing network request w/ NOT_FOUND error', async function() {
+    // GIVEN a XpringClient which wraps an erroring network client that returns an Error containing
+    // a property 'code' with value grpc.status.NOT_FOUND.
+    const errorWithCode5 = new FakeServiceError(
```
Comment (keefertaylor) [inline test/default-xpring-client-test.ts]: I would call this `notFoundError`, since `5` is kind of a magic number and this isn't a really descriptive name.

File: test/default-xpring-client-test.ts [hunk 10]
```diff
@@ -436,17 +454,52 @@ describe('Default Xpring Client', function(): void {
...
     const exists = await xpringClient.accountExists(testAddress)
 
     // THEN accountExists returns false
     assert.equal(exists, false)
   })
 
+  it('Check if account exists - failing network request w/ CANCELLED error', function(done) {
+    // GIVEN a XpringClient which wraps an erroring network client that returns an Error containing
+    // a property 'code' with value grpc.status.CANCELLED.
+    const errorWithCode1 = new FakeServiceError(
```
Comment (keefertaylor) [inline test/default-xpring-client-test.ts]: s/errorWithCode1/cancelledError

File: test/integration-test.ts [hunk 11]
```diff
@@ -136,10 +136,25 @@ describe('Xpring JS Integration Tests', function(): void {
...
+    void
+  > {
     this.timeout(timeoutMs)
 
     const doesExist = await xpringClient.accountExists(recipientAddress)
+
     assert.equal(doesExist, true)
   })
+
+  // Omitting false-case tests for LegacyNode and LegacyWeb
```
Comment (keefertaylor) [inline test/integration-test.ts]: Can we be slightly more descriptive here? 

// Tests for account existence on the legacy shim are intentionally omitted as the legacy shim does have the required granularity of information.

File: test/legacy/legacy-default-xpring-client-test.ts [hunk 12]
```diff
@@ -35,6 +35,23 @@ const transactionStatusFailureCodes = [
 
 const transactionHash = 'DEADBEEF'
 
+/**
```
Comment (keefertaylor) [inline test/legacy/legacy-default-xpring-client-test.ts]: Definitely don't write this class twice. 

You could just make it's own file in `tests/fakes/fake-grpc-error.ts`) or stick it in `fake-network-client`. I'd probably recommend the former as `fake-network-client` is already becoming pretty long, but don't have strong opinions if you want to keep all the fake network stuff as one logical unit.

**Comment by amiecorso** [comment]
> I changed the branch this PR is compared to as I see now that #182 is what you were building on top of which is where those mocha changes came in.

File: test/default-xpring-client-test.ts (hunk 6 above)
Comment (amiecorso) [reply test/default-xpring-client-test.ts]: agreed!

File: test/default-xpring-client-test.ts (hunk 9 above)
Comment (amiecorso) [reply test/default-xpring-client-test.ts]: way better

File: test/legacy/legacy-default-xpring-client-test.ts (hunk 12 above)
Comment (amiecorso) [reply test/legacy/legacy-default-xpring-client-test.ts]: great, this is in its own file now.

File: test/integration-test.ts (hunk 11 above)
Comment (amiecorso) [reply test/integration-test.ts]: updated

File: test/default-xpring-client-test.ts (hunk 8 above)
Comment (amiecorso) [reply test/default-xpring-client-test.ts]: this makes sense too - less reference to actual names of things, and more of an abstract conception of what's happening

File: test/fakes/fake-grpc-error.ts [hunk 13]
```diff
@@ -0,0 +1,18 @@
+import grpc from 'grpc'
+/**
```
Comment (keefertaylor) [inline test/fakes/fake-grpc-error.ts]: nit: one vertical blank line after imports

File: test/fakes/fake-grpc-error.ts [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+ * Convenience class for creating more specific Error objects to mimick grpc error responses.
```
Comment (keefertaylor) [inline test/fakes/fake-grpc-error.ts]: also super nit: `gRPC`

File: src/default-xpring-client.ts [hunk 15]
```diff
@@ -299,7 +300,13 @@ class DefaultXpringClient implements XpringClientDecorator {
       await this.getBalance(address)
       return true
     } catch (e) {
-      return false
+      if (
+        Object.prototype.hasOwnProperty.call(e, 'code') &&
```
Comment (keefertaylor) [inline src/default-xpring-client.ts]: I'm surprised you can't use typescripts optional chaining here:
```
if e.code? === grpc.status.NOT_FOUND
```

If code is `undefined` we'll safely unwrap it to undefined and the check will fail. Otherwise, it'll be unwrapped to a status code and compared.

File: test/fakes/fake-grpc-error.ts (hunk 14 above)
Comment (amiecorso) [reply test/fakes/fake-grpc-error.ts]: does this mean we want to ```import grpc from 'gRPC'``` ?

Comment (amiecorso) [reply test/fakes/fake-grpc-error.ts]: oh you must mean gRPC in the comment, because it doesn't know what gRPC is as an import

File: src/default-xpring-client.ts (hunk 15 above)
Comment (amiecorso) [reply src/default-xpring-client.ts]: So I'm pretty sure we actually can use this construct.  I'll double-check with typescript people here to make sure nothing stupid happens, but if they approve I'll change that.

Comment (amiecorso) [reply src/default-xpring-client.ts]: so the optional chaining syntax doesn't seem to work in that particular way, but I updated the if statement so it's a bit shorter.
